src/SubPoolFactory.sol

```diff
@@ -0,0 +1,196 @@
+    event UpdateMinCowAmt(uint256 amt);
+    event UpdateBackendUri(address indexed pool, string uri);
+    event SolverPoolDeployed(address indexed solver, address indexed pool);
+    event SolverPoolBilled(address indexed pool, uint256 amt, uint256 cowAmt, uint256 ethAmt, string reason);
+    event AnnounceExit(address indexed pool);
+    event Exit(address indexed pool);
+
+    struct SubPoolData {
+        /// The collateral token that pool was initialized with.
+        address collateral;
+        /// The timestamp at which the pool can exit.
+        uint88 exitTimestamp;
+        /// Whether the pool has exited.
+        bool hasExited;
+    }
+
+    struct Config {
+        /// the delay between `announceExit` and `exit`.
+        uint32 exitDelay;
+        /// minimum amount of COW to be bonded.
+        uint224 minCowAmt;
+    }
+
+    /// @notice Individual subpool datas
+    mapping(address => SubPoolData) public subPoolData;
+    /// @notice Solver backend uri
+    /// @dev    Not stored in the SubPoolData because we don't want to load the string
+    ///         everytime we read the subpooldata in memory.
+    mapping(address => string) public backendUri;
+    /// @notice system cfg
+    Config public cfg;
+    /// @notice COW token
+    address public immutable COW;
+
+    constructor(uint32 exitDelay, uint224 minCowAmt, address cow) {
+        cfg = Config({exitDelay: exitDelay, minCowAmt: minCowAmt});
+        emit UpdateExitDelay(exitDelay);
+        emit UpdateMinCowAmt(minCowAmt);
+        COW = cow;
+        _addOwner(msg.sender);
+    }
+
+    /// @notice Set the delay between `quit` and `exit`.
+    function setExitDelay(uint32 delay) external auth {
+        cfg.exitDelay = delay;
+        emit UpdateExitDelay(delay);
+    }
+
+    /// @notice Set the min cow amount.
+    function setMinCowAmt(uint224 amt) external auth {
+        cfg.minCowAmt = amt;
+        emit UpdateMinCowAmt(amt);
+    }
+
+    /// @notice Create a `SubPool` for the user at a deterministic address with salt as `msg.sender`.
+    /// @param token  - The token to use as collateral.
+    function create(address token, uint256 amt, uint256 cowAmt, string calldata uri) external returns (address) {
+        SubPool subpool = new SubPool{salt: bytes32(uint256(uint160(msg.sender)))}(msg.sender, COW);
+        subpool.initializeCollateralToken(token);
+        emit SolverPoolDeployed(msg.sender, address(subpool));
+        Config memory config = cfg;
+
+        if (cowAmt < config.minCowAmt) {
+            revert SubPoolFactory__InsufficientCollateral();
+        }
+
+        SafeTransferLib.safeTransferFrom(token, msg.sender, address(subpool), amt);
+        SafeTransferLib.safeTransferFrom(COW, msg.sender, address(subpool), cowAmt);
+
+        subPoolData[address(subpool)] = SubPoolData({collateral: token, exitTimestamp: 0, hasExited: false});
+        backendUri[address(subpool)] = uri;
+        emit UpdateBackendUri(address(subpool), uri);
```

Comment:
To minimize reentrancy risk, I'd order these lines as follows:
```suggestion
        subPoolData[address(subpool)] = SubPoolData({collateral: token, exitTimestamp: 0, hasExited: false});
        backendUri[address(subpool)] = uri;
        emit UpdateBackendUri(address(subpool), uri);

        SafeTransferLib.safeTransferFrom(COW, msg.sender, address(subpool), cowAmt);
        SafeTransferLib.safeTransferFrom(token, msg.sender, address(subpool), amt);
```

In this way, the only untrusted call is the last, and since there would be no state change after the untrusted call then we don't have to worry about the fact that a bad `token` might be using the subpool while it's not fully initialized yet.